### ISSUE TITLE
P1 experimental melody: integrate MidiBERT-Piano as an explicit runtime challenger

### ISSUE THREAD

**Issue by gr-rr** [ListenCloser/listencloser#1233]
Parent product/UX: #1194 #1173. Historical evaluation/provenance: #1026. Representation authority: #613. Representation quality: #498.

## Product problem

The Melody object is now useful enough to expose model quality directly: the user can see, inspect, and hear the proposed melody, but the current `lstom` interpretation is often musically weak outside its narrow validated domain.

The UI/product object does **not** need to change for the next quality step. The next bounded experiment is to make a stronger symbolic melody model runnable behind the existing `MelodyEngine` contract so real product use can compare interpretations before a canonization decision.

## Current runtime truth

Current `main` already has the correct replaceable boundary:

```text
performance MIDI
  -> MelodyEngine.analyze(midi_bytes)
  -> normalized MelodyResult
  -> existing persisted melody note evidence
  -> existing Melody lane / Focus / Hear / Piano Roll linkage
```

`backend/engines/registry.py` currently exposes only `lstom` and `skyline`; `MELODY_ENGINE=lstom` remains the default.

Historical #1026 established MidiBERT-Piano as a credible challenger but stopped at `RESEARCH / REVISIT` because the old POP909 LStoM holdout/checkpoint provenance could not support a leakage-safe same-split comparison. That historical evaluation blocker should **not** prevent a reversible experimental runtime under the current roadmap policy (`integrate -> inspect -> evaluate when a durable default decision exists`).

## Candidate

**MidiBERT-Piano CP melody checkpoint** from `wazenmai/MIDI-BERT`.

Current upstream facts to verify/pin in implementation:
- repository code is MIT;
- upstream explicitly provides MIDI -> predicted-melody inference and a public melody checkpoint;
- CPU evaluation/inference is supported;
- training/task domain is POP909-style symbolic piano/pop, so classical/general claims remain unvalidated;
- current CP branch head observed during issue creation: `0b935584f641d3d59a8e6aff2f334b425ce1542d`;
- checkpoint/model artifact usage/redistribution terms must be verified separately from repository code before deployed exposure.

## First runtime slice

Add one bounded `MidiBERTMelodyEngine` that implements the existing `MelodyEngine` protocol and emits the same normalized product contract consumed by the current Melody UI.

Required behavior:

1. accept the exact source MIDI bytes already used by Melody;
2. run the official CP melody inference semantics with a pinned source revision/checkpoint identity;
3. map predicted melody notes back to exact source MIDI note/time evidence without inventing a second coordinate system;
4. preserve native method semantics (including any melody/bridge/accompaniment class mapping) explicitly rather than silently changing labels;
5. emit exact engine/source/checkpoint/preprocessing provenance and checkpoint checksum;
6. support explicit selection (`name="midibert"` / equivalent request-level engine identity) while **keeping `lstom` as the default**;
7. selected-engine failure is explicit; no silent fallback to LStoM;
8. existing Melody lane / Hear / Focus / Piano Roll linkage should consume the result without a new top-level UI or new result model.

## Dependency policy

Do **not** install the upstream 2021 `requirements.txt` wholesale into the production worker. It pins/requires an old research stack (`mido 1.2.10`, `scikit-learn 0.24.2`, `transformers 4.8.2`, old torchaudio assumptions) while ListenCloser already owns Python 3.11 + Torch 2.6 worker compatibility.

Prefer the smallest lawful inference port/adapter against the current worker stack:
- reuse only the minimum upstream CP tokenization/inference code needed for parity;
- keep heavy imports lazy;
- do not add a second generic model/plugin framework;
- if exact inference cannot be made compatible without a large dependency fork or algorithm reimplementation, stop and record the blocker rather than contaminating the worker environment.

## Product proof now; benchmark later

The goal of this issue is **runtime availability + real product inspection**, not immediate canonization.

On a small set of lawful Works, including at least:
- one pop/arranged piano-like positive case;
- one current real Melody failure;
- one classical/contrapuntal negative or out-of-domain case;

record:
- whether the result is visibly/hearably more useful than current LStoM;
- obvious omission/contamination/failure behavior;
- CPU wall time, cold/warm load, and peak RSS on the normal worker class;
- exact source/checkpoint identity.

Do not claim superiority from this smoke. A later focused evaluation is justified only if real product use creates a default/promotion decision.

## Acceptance

- [ ] `MidiBERTMelodyEngine` runs on a real product MIDI through the existing Melody engine boundary;
- [ ] existing Melody output/UI consumes it without a new product concept;
- [ ] exact selected source-note/time mapping and Version provenance are preserved;
- [ ] source revision + checkpoint checksum + artifact terms are recorded;
- [ ] LStoM remains default; MidiBERT is explicit/experimental only;
- [ ] failure never silently substitutes another engine;
- [ ] no wholesale legacy dependency stack or generic engine framework is introduced;
- [ ] several real Works are manually inspected and runtime/RSS recorded;
- [ ] capability metadata describes MidiBERT as experimental/domain-limited, not canonical;
- [ ] canonical CI/real-stack gates relevant to the changed path are green.

## Stop conditions

Stop rather than force integration if:
- checkpoint terms do not permit the intended experimental deployment mode;
- exact note mapping cannot preserve current source-Version semantics;
- the upstream inference contract requires a materially incompatible runtime/dependency fork that is larger than a bounded adapter.

No broad POP909 benchmark is required before the first runtime slice.

Refs #1026 #1194 #1173 #613 #498 #1218.

**Comment by gr-rr** [comment]
### Product-owner refinement — include the thin alternate-interpretation UI

This runtime slice should not stop at API-only selection if the checkpoint/runtime gate clears. Reuse #1173's existing grammar on the **already-shipped Melody object**:

```text
Melody · Experimental
  Details
    Interpretation · LStoM
    Try another interpretation
      MidiBERT-Piano · Experimental
```

Requirements:
- LStoM remains the current/default interpretation;
- generating MidiBERT creates/preserves an independent result rather than overwriting LStoM;
- switching an already-generated interpretation is cheap/reversible and preserves shared playhead/Focus/Piano Roll note linkage;
- engine/model names stay under Details, not as always-visible Melody chrome;
- no new Melody tab, model matrix, comparison dashboard, or first-use choice;
- the user can immediately **Hear** the selected melody interpretation using the existing reduction playback path.

The product object stays identical; only the method behind it changes. This is the intended proof of the stable-UX / replaceable-engine architecture.
